src/shell/jobspec.c

```diff
@@ -0,0 +1,160 @@
+/************************************************************\
+ * Copyright 2019 Lawrence Livermore National Security, LLC
+ * (c.f. AUTHORS, NOTICE.LLNS, COPYING)
+ *
+ * This file is part of the Flux resource manager framework.
+ * For details, see https://github.com/flux-framework.
+ *
+ * SPDX-License-Identifier: LGPL-3.0
+\************************************************************/
+
+#if HAVE_CONFIG_H
+#include "config.h"
+#endif
+#include <flux/core.h>
+#include <jansson.h>
+
+#include "src/common/libutil/log.h"
+
+#include "jobspec.h"
+
+struct res_level {
+    const char *type;
+    int count;
+    json_t *with;
+};
+
+static void set_error (json_error_t *error, const char *fmt, ...)
+{
+    va_list ap;
+
+    if (error) {
+        va_start (ap, fmt);
+        vsnprintf (error->text, sizeof (error->text), fmt, ap);
+        va_end (ap);
+    }
+}
+
+static int parse_res_level (json_t *o,
+                            int level,
+                            struct res_level *resp,
+                            json_error_t *error)
+{
+    json_error_t loc_error;
+    struct res_level res;
+
+    if (o == NULL) {
+        set_error (error, "level %d: missing", level);
+        return -1;
+    }
+    res.with = NULL;
+    /* For jobspec version 1, expect exactly one array element per level.
+     */
+    if (json_unpack_ex (o, &loc_error, 0,
+                        "[{s:s s:i s?o}]",
```

Comment:
what is `?` for.  Skimming API ref, `?` is only for packing not unpacking?

Reply:
nevermind, I now see: `New in version 2.3: Any s representing a key may be suffixed with a ? to make the key optional. If the key is not found, nothing is extracted. See below for an example.`

Reply:
From https://jansson.readthedocs.io/en/2.6/apiref.html#parsing-and-validating-values

> New in version 2.3: Any s representing a key may be suffixed with a ? to make the key optional. If the key is not found, nothing is extracted. See below for an example.